### PR TITLE
Add `UnwindSafe` to `Once`

### LINKED ISSUE #43469: std::sync::Once: Should Once be UnwindSafe or RefUnwindSafe?
Currently, the following program [will not compile](https://play.rust-lang.org/?gist=b2be5c6728418ded9d6a559c780779b7&version=stable) because `Once` isn't `RefUnwindSafe`. Is this the desired behavior, or should we consider making it `RefUnwindSafe` or `UnwindSafe`?

```rust
use std::sync::Once;
use std::panic::catch_unwind;

fn main() {
    let o = Once::new();
    catch_unwind(|| o.call_once(|| panic!("foo")));
}
```

### PR DESCRIPTION
**Pull request by Milo123459**
Fixes #43469

### CODE REVIEW

**Comment by camelid** [comment]
r? rust-lang/libs

**Comment by camelid** [comment]
r? rust-lang/libs-api

**Review by dtolnay — CHANGES_REQUESTED** [review]
This needs a manual `RefUnwindSafe` impl as well.

**Comment by dtolnay** [comment]
@rust-lang/libs-api:
@rfcbot fcp merge

All of the `Once` methods behave atomically and there is no possibility of the `Once` being left in a corrupted state following an unwind. `UnwindSafe` and `RefUnwindSafe` auto trait impls only didn't kick in because of the presence of a `Cell` inside `Waiter`.

File: library/std/src/sync/once.rs
Comment (camelid) [inline library/std/src/sync/once.rs:127]: (Version will need to be updated.)

Comment (camelid) [reply library/std/src/sync/once.rs:127]: Also, should this use a new feature—like `sync_once_ref_unwind_safe`—to avoid confusion?

File: library/std/src/sync/once.rs
Comment (camelid) [inline library/std/src/sync/once.rs:127]: This actually needs to be at least 1.59.0, since that is the current nightly version. It may be better to wait until this is ready to be merged to update the version though.

```suggestion
#[stable(feature = "sync_once_ref_unwind_safe", since = "1.59.0")]
```

**Comment by dtolnay** [comment]
- Squashed, and added `RefUnwindSafe` as requested above (and implied by the branch name)

**Review by dtolnay — APPROVED** [review]

**Pull request merged**
